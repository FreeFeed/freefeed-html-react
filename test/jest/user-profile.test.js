import React from 'react';
import { shallow } from 'enzyme';

import UserProfile from '../../src/components/user-profile';

const renderUserProfile = (props = {}) => {
  const defaultProps = {
    id: 'user-id',
    type: 'user',
    username: 'username',
    screenName: 'ScreenName',
    profilePictureLargeUrl: 'profilePictureLargeUrl',
    description: 'description',
    amIGroupAdmin: false,
    isLoading: false,
    isUserFound: true,
    canISeeSubsList: true,
    statistics: {
      subscriptions: 10,
      subscribers: 20,
      comments: 30,
      likes: 40,
    },
    authenticated: true,
    isItMe: false,
    blocked: false,
    isPrivate: false,
    subscribed: true,
    hasRequestBeenSent: false,
    userView: {
      isSubscribing: false,
    },
    canAcceptDirects: true,
    canIPostHere: true,
    isRestricted: false,
    createPostViewState: () => {},
    sendTo: () => {},
    user: 'user',
    sendSubscriptionRequest: () => {},
    subscribe: () => {},
    unsubscribe: () => {},
    ban: () => {},
    createPost: () => {},
    resetPostCreateForm: () => {},
    expandSendTo: () => {},
    addAttachmentResponse: () => {},
  };

  return shallow(<UserProfile {...defaultProps} {...props} />);
};

describe('<UserProfile>', () => {
  it('Should render with default data', () => {
    const actual = renderUserProfile();
    expect(actual).toMatchSnapshot();
  });

  it('Should render when is a group', () => {
    const actual = renderUserProfile({
      type: 'group',
      amIGroupAdmin: true,
    });
    expect(actual).toMatchSnapshot();
  });

  it('Should render 404 when user is not found', () => {
    const actual = renderUserProfile({ isLoading: false, isUserFound: false });
    expect(actual).toMatchSnapshot();
  });

  it('Should render when private and not subscribed to', () => {
    const actual = renderUserProfile({ isPrivate: '1', subscribed: false });
    expect(actual).toMatchSnapshot();
  });

  it('Should not see links to subscription lists when canISeeSubsList is false', () => {
    const actual = renderUserProfile({ canISeeSubsList: false });
    expect(actual).toMatchSnapshot();
  });

  it('Should not show stats when user is blocked', () => {
    const actual = renderUserProfile({ blocked: true });
    expect(actual).toMatchSnapshot();
  });

  it('Should render a warning for a restricted group', () => {
    const actual = renderUserProfile({ type: 'group', canIPostHere: false, isRestricted: '1' });
    expect(actual).toMatchSnapshot();
  });

  it('Should block when block button is clicked', () => {
    const ban = jest.fn();
    const actual = renderUserProfile({ subscribed: false, ban });
    const blockButton = actual.find('.profile-action-block');
    const fakeClickEvent = { preventDefault: () => {} };

    blockButton.simulate('click', fakeClickEvent);
    expect(ban).toHaveBeenCalledTimes(1);
    expect(ban).toHaveBeenCalledWith({
      id: 'user-id',
      username: 'username',
    });
  });

  it('Should subscribe when subscribe button is clicked', () => {
    const subscribe = jest.fn();
    const actual = renderUserProfile({ subscribed: false, subscribe });
    const subscribeButton = actual.find('.subscribe-control-subscribe');
    const fakeClickEvent = { preventDefault: () => {} };

    subscribeButton.simulate('click', fakeClickEvent);
    expect(subscribe).toHaveBeenCalledTimes(1);
    expect(subscribe).toHaveBeenCalledWith({
      id: 'user-id',
      username: 'username',
    });
  });

  it('Should unsubscribe when unsubscribe button is clicked', () => {
    const unsubscribe = jest.fn();
    const actual = renderUserProfile({ subscribed: true, unsubscribe });
    const unsubscribeButton = actual.find('.subscribe-control-unsubscribe');
    const fakeClickEvent = { preventDefault: () => {} };

    unsubscribeButton.simulate('click', fakeClickEvent);
    expect(unsubscribe).toHaveBeenCalledTimes(1);
    expect(unsubscribe).toHaveBeenCalledWith({
      id: 'user-id',
      username: 'username',
    });
  });

  it('Should send a subscribe request when requesting subscription to a private user', () => {
    const sendSubscriptionRequest = jest.fn();
    const actual = renderUserProfile({
      subscribed: false,
      isPrivate: '1',
      sendSubscriptionRequest,
    });
    const subscriptionRequestButton = actual.find('.subscribe-control-request-subscription');
    const fakeClickEvent = { preventDefault: () => {} };

    subscriptionRequestButton.simulate('click', fakeClickEvent);
    expect(sendSubscriptionRequest).toHaveBeenCalledTimes(1);
    expect(sendSubscriptionRequest).toHaveBeenCalledWith({
      id: 'user-id',
      username: 'username',
    });
  });
});
