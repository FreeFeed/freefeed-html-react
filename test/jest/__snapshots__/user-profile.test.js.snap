// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile> Should not see links to subscription lists when canISeeSubsList is false 1`] = `
<div>
  <ErrorBoundary>
    <div
      className="profile"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-9 col-xs-12"
        >
          <div
            className="avatar"
          >
            <img
              height="75"
              src="profilePictureLargeUrl"
              width="75"
            />
          </div>
          <div
            className="description"
          >
            <div
              className="name"
              dir="auto"
            >
              ScreenName
            </div>
            <PieceOfText
              isExpanded={true}
              text="description"
            />
          </div>
        </div>
        <div
          className="col-sm-3 col-xs-12"
        >
          <div
            className="profile-stats"
          >
            <div
              className="profile-stats-item"
            >
              10 subscriptions
            </div>
             
            <div
              className="profile-stats-item"
            >
              20 subscribers
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/comments"
              >
                30 comments
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/likes"
              >
                40 likes
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="profile-controls"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-7 col-sm-7 subscribe-controls"
        >
          <a
            className="subscribe-control-unsubscribe"
            onClick={[Function]}
          >
            Unsubscribe
          </a>
        </div>
        <div
          className="col-xs-5 col-sm-5 text-right"
        >
          <ul
            className="profile-actions"
          >
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?to=username"
              >
                Direct message
              </Link>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <CreatePost
      addAttachmentResponse={[Function]}
      createPost={[Function]}
      createPostViewState={[Function]}
      expandSendTo={[Function]}
      resetPostCreateForm={[Function]}
      sendTo={[Function]}
      user="user"
    />
  </ErrorBoundary>
</div>
`;

exports[`<UserProfile> Should not show stats when user is blocked 1`] = `
<div>
  <ErrorBoundary>
    <div
      className="profile"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-9 col-xs-12"
        >
          <div
            className="avatar"
          >
            <img
              height="75"
              src="profilePictureLargeUrl"
              width="75"
            />
          </div>
          <div
            className="description"
          >
            <div
              className="name"
              dir="auto"
            >
              ScreenName
            </div>
            <PieceOfText
              isExpanded={true}
              text="description"
            />
          </div>
        </div>
      </div>
    </div>
    <CreatePost
      addAttachmentResponse={[Function]}
      createPost={[Function]}
      createPostViewState={[Function]}
      expandSendTo={[Function]}
      resetPostCreateForm={[Function]}
      sendTo={[Function]}
      user="user"
    />
  </ErrorBoundary>
</div>
`;

exports[`<UserProfile> Should render 404 when user is not found 1`] = `
<div>
  <ErrorBoundary>
    <h2>
      404 Not Found
    </h2>
    <CreatePost
      addAttachmentResponse={[Function]}
      createPost={[Function]}
      createPostViewState={[Function]}
      expandSendTo={[Function]}
      resetPostCreateForm={[Function]}
      sendTo={[Function]}
      user="user"
    />
  </ErrorBoundary>
</div>
`;

exports[`<UserProfile> Should render a warning for a restricted group 1`] = `
<div>
  <ErrorBoundary>
    <div
      className="profile"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-9 col-xs-12"
        >
          <div
            className="avatar"
          >
            <img
              height="75"
              src="profilePictureLargeUrl"
              width="75"
            />
          </div>
          <div
            className="description"
          >
            <div
              className="name"
              dir="auto"
            >
              ScreenName
            </div>
            <PieceOfText
              isExpanded={true}
              text="description"
            />
          </div>
        </div>
        <div
          className="col-sm-3 col-xs-12"
        >
          <div
            className="profile-stats"
          >
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/subscribers"
              >
                20 subscribers
              </Link>
            </div>
             
             
          </div>
        </div>
      </div>
    </div>
    <div
      className="profile-controls"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-7 col-sm-7 subscribe-controls"
        >
          <a
            className="subscribe-control-unsubscribe"
            onClick={[Function]}
          >
            Unsubscribe
          </a>
        </div>
        <div
          className="col-xs-5 col-sm-5 text-right"
        >
          <ul
            className="profile-actions"
          >
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?to=username"
              >
                Direct message
              </Link>
            </li>
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?invite=username"
              >
                Invite
              </Link>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="create-post create-post-restricted"
    >
      Only administrators can post to this group.
    </div>
  </ErrorBoundary>
</div>
`;

exports[`<UserProfile> Should render when is a group 1`] = `
<div>
  <ErrorBoundary>
    <div
      className="profile"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-9 col-xs-12"
        >
          <div
            className="avatar"
          >
            <img
              height="75"
              src="profilePictureLargeUrl"
              width="75"
            />
          </div>
          <div
            className="description"
          >
            <div
              className="name"
              dir="auto"
            >
              ScreenName
            </div>
            <PieceOfText
              isExpanded={true}
              text="description"
            />
          </div>
        </div>
        <div
          className="col-sm-3 col-xs-12"
        >
          <div
            className="profile-stats"
          >
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/subscribers"
              >
                20 subscribers
              </Link>
            </div>
             
             
          </div>
        </div>
      </div>
    </div>
    <div
      className="profile-controls"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-7 col-sm-7 subscribe-controls"
        >
          <a
            className="subscribe-control-unsubscribe"
            onClick={[Function]}
          >
            Unsubscribe
          </a>
        </div>
        <div
          className="col-xs-5 col-sm-5 text-right"
        >
          <ul
            className="profile-actions"
          >
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?to=username"
              >
                Direct message
              </Link>
            </li>
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?invite=username"
              >
                Invite
              </Link>
            </li>
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/settings"
              >
                Settings
              </Link>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <CreatePost
      addAttachmentResponse={[Function]}
      createPost={[Function]}
      createPostViewState={[Function]}
      expandSendTo={[Function]}
      resetPostCreateForm={[Function]}
      sendTo={[Function]}
      user="user"
    />
  </ErrorBoundary>
</div>
`;

exports[`<UserProfile> Should render when private and not subscribed to 1`] = `
<div>
  <ErrorBoundary>
    <div
      className="profile"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-9 col-xs-12"
        >
          <div
            className="avatar"
          >
            <img
              height="75"
              src="profilePictureLargeUrl"
              width="75"
            />
          </div>
          <div
            className="description"
          >
            <div
              className="name"
              dir="auto"
            >
              ScreenName
            </div>
            <PieceOfText
              isExpanded={true}
              text="description"
            />
          </div>
        </div>
        <div
          className="col-sm-3 col-xs-12"
        >
          <div
            className="profile-stats"
          >
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/subscriptions"
              >
                10 subscriptions
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/subscribers"
              >
                20 subscribers
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/comments"
              >
                30 comments
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/likes"
              >
                40 likes
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="profile-controls"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-7 col-sm-7 subscribe-controls"
        >
          <a
            className="subscribe-control-request-subscription"
            onClick={[Function]}
          >
            Request a subscription
          </a>
        </div>
        <div
          className="col-xs-5 col-sm-5 text-right"
        >
          <ul
            className="profile-actions"
          >
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?to=username"
              >
                Direct message
              </Link>
            </li>
            <li>
              <a
                className="profile-action-block"
                onClick={[Function]}
              >
                Block this user
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <CreatePost
      addAttachmentResponse={[Function]}
      createPost={[Function]}
      createPostViewState={[Function]}
      expandSendTo={[Function]}
      resetPostCreateForm={[Function]}
      sendTo={[Function]}
      user="user"
    />
  </ErrorBoundary>
</div>
`;

exports[`<UserProfile> Should render with default data 1`] = `
<div>
  <ErrorBoundary>
    <div
      className="profile"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-9 col-xs-12"
        >
          <div
            className="avatar"
          >
            <img
              height="75"
              src="profilePictureLargeUrl"
              width="75"
            />
          </div>
          <div
            className="description"
          >
            <div
              className="name"
              dir="auto"
            >
              ScreenName
            </div>
            <PieceOfText
              isExpanded={true}
              text="description"
            />
          </div>
        </div>
        <div
          className="col-sm-3 col-xs-12"
        >
          <div
            className="profile-stats"
          >
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/subscriptions"
              >
                10 subscriptions
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/subscribers"
              >
                20 subscribers
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/comments"
              >
                30 comments
              </Link>
            </div>
             
            <div
              className="profile-stats-item"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/username/likes"
              >
                40 likes
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="profile-controls"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-7 col-sm-7 subscribe-controls"
        >
          <a
            className="subscribe-control-unsubscribe"
            onClick={[Function]}
          >
            Unsubscribe
          </a>
        </div>
        <div
          className="col-xs-5 col-sm-5 text-right"
        >
          <ul
            className="profile-actions"
          >
            <li>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/filter/direct?to=username"
              >
                Direct message
              </Link>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <CreatePost
      addAttachmentResponse={[Function]}
      createPost={[Function]}
      createPostViewState={[Function]}
      expandSendTo={[Function]}
      resetPostCreateForm={[Function]}
      sendTo={[Function]}
      user="user"
    />
  </ErrorBoundary>
</div>
`;
